### ISSUE TITLE
Add Task.Parallel to allow running tasks concurrently

### ISSUE THREAD

**Issue by chrnorm** [goyek/goyek#179]
## Why

I'm looking at `goyek` as an alternative to Mage - thanks for open sourcing this, so far I am loving the API as it feels very similar to the built in `testing` library and to libraries like `urfav/cli`.

Something that I liked in Mage is that it was very easy to run tasks concurrently. For example, if task `all` depends on tasks `a` and `b`, tasks `a` and `b` would be executed concurrently in separate Goroutines. I tried modelling this in goyek as follows:

```go
package main

import (
	"time"

	"github.com/goyek/goyek"
)

func main() {
	flow := &goyek.Flow{}

	a := flow.Register(goyek.Task{
		Name: "a",
		Action: func(tf *goyek.TF) {
			time.Sleep(time.Second * 1)
			tf.Log("a")
			time.Sleep(time.Second * 1)
			tf.Log("a")
			time.Sleep(time.Second * 1)
			tf.Log("a")
		},
	})

	b := flow.Register(goyek.Task{
		Name: "b",
		Action: func(tf *goyek.TF) {
			time.Sleep(time.Second * 1)
			tf.Log("b")
			time.Sleep(time.Second * 1)
			tf.Log("b")
			time.Sleep(time.Second * 1)
			tf.Log("b")
		},
	})

	flow.Register(goyek.Task{
		Name: "all",
		Deps: goyek.Deps{a, b},
	})

	flow.Main()
}
```

It appears that the tasks run sequentially, even though `a` and `b` don't depend on one another.

```
go run ./build all -v
===== TASK  a
a
a
a
----- PASS: a (3.00s)
===== TASK  b
b
b
b
----- PASS: b (3.00s)
ok      6.006s
```

## What

Would you be open to a contribution to support running tasks concurrently? 

Some approaches to the API could be to add a flag at the Flow level:

```go
flow := &goyek.Flow{
	Concurrent: true,
}

or to allow concurrency to be specified at a task level:

```go
flow.Register(goyek.Task{
	Name: "all",
	Deps: goyek.ConcurrentDeps{a, b},
})
```

**Comment by pellared** [comment]
> Would you be open to a contribution to support running tasks concurrently?

Yes, I am open both to your contribution as well adding this functionality to `goyek`👍 🚀 

Regarding the API, I propose to make add a `tf.Parallel()` method that would work like [`t.Parallel`](https://pkg.go.dev/testing#T.Parallel).

1. I think it is usually the nature of given action (task) that it can run concurrently.
2. It should work well combined with https://github.com/goyek/goyek/issues/116 (assuming it will be implemented at some point)
3. As you even said - having API similar to the `testing` package has its benefits 😉

Regarding the behavior, I think it would be important to buffer the output passed to `tf.Output` so that the output of the actions running in parallel is not mixed.

Regarding the implementation, it may be useful to use `golang.org/x/sync/errgroup` for running the tasks that may run concurrently.

Regarding testing, I think we would need a crazy test that mixes sequential and parallel tasks and some mix of dependencies. The test could verify the log output if the order is "correct" which means
1. The order is OK (dependencies are respected)
2. The tasks' output is not "cut"
Moreover, we would need a test that checks the behavior when an action running concurrently fails.

WDYT?

**Comment by pellared** [comment]
PS. Thanks for sharing `urfave/cli` - I was not aware of it and it looks awesome 👍

**Comment by pellared** [comment]
This may be helpful: https://engineering.mercari.com/en/blog/entry/20220408-how_to_use_t_parallel/

**Comment by pellared** [comment]
@chrnorm Could you provide me some real-world examples of when this feature would be very helpful? I am working on v2 of goyek and I have this feature back in my mind.  Personally, I never needed such a feature.

For all my build pipelines, it was best to run all tasks sequentially as usually each of them was usually internally parallelized. Commands like `go build`, `go test`, `golangci-lint` take the advantage of many cores.

Moreover running the dependencies sequentially makes the UX better as the output can be streamed without "mixing" the tasks' outputs. 

Also, you get a chance that individual dependencies to fail faster (as they have more resources to run faster). Therefore, I tend to order the short-running dependencies before the long-running ones. Such optimization would not be possible if they would run concurrently as we would have to wait until all (sic!) concurrent dependencies are finished (for proper cleanup).

At last, I do not know at which level something should be declared that it can be run concurrently Should it be on the dependency level (e.g. by setting `Task.ConcurrentDeps`)? On the task level (e.g. by setting `Task.Concurrent = true` or running `tf.Parallel()` inside the task's action)?

@natefinch could you please try to enlighten me as the author of Mage? 😉

**Comment by natefinch** [comment]
I actually assume most jobs run single threaded. Aside from some highly optimized go programs, I think most jobs do one thing and do it on a single thread. Like copying a file or querying an API. Almost all CLI commands are single threaded.  IMO the overhead of switching threads and contention between running processes is unlikely to overshadow the benefits you get from not having to wait for a bunch of single threaded things to run.  

Of course, there are times when your subprocesses *are* heavy and multi threaded, in which case it may be a good idea to have a way to ensure they're run sequentially, but I think that's generally the exception, not the rule.

**Comment by pellared** [comment]
Closing as "won't fix" until there are further requests or proposals. Also the issue author is not responsive.

**Comment by pellared** [comment]
Reopening, but if anyone from the community would like to implement it.

**Comment by pellared** [comment]
It may be simpler to mark the task as being able to run in parallel by adding a `Parallel bool` field to `Task` type.

**Comment by jrossi** [comment]
I don't know if this is the same things or something completely different, but my use-case would make use of parellel like features. 

My use case is to depend up task, but that task continues running, but also recevives a cancel to it's context when all depentant tasks are completed.    This allow me to do a full intergration environment using goyek that has databases, redis, etc all started and running when before other tasks run. 

```go
var build = // Build code 
var db_run  = goyek.Define(goyek.Task{
	Name:  "db_run",
	Usage: "",
        Background: true, 
	Deps: goyek.Deps{
		build,
	},
	Action: func(a *goyek.A) {
	       findCmd := cmd.NewCmd("db start")
	       statusChan := findCmd.Start() // non-blocking     
	       select {
	       case finalStatus := <-statusChan:
	             // Error something happened with before expected exit
	       case <-a.Context.Done():
                     // All depent have completed
                     // close db
                     findCmd.Close()
                     return  
	      }
	}
}
var db_migrate  = goyek.Define(goyek.Task{
	Name:  "db_migrate",
	Usage: "",
        Background: false, 
	Deps: goyek.Deps{
		build,
		db_run,
	},
	Action: func(a *goyek.A) {
               cmd.Exec("load_db migrations/*.sql")
	}
}
```

**Comment by pellared** [comment]
Proposing `A.Parallel` method. As it would work nicely with `A.Run` method. See: https://github.com/goyek/goyek/issues/116

**Comment by pellared** [comment]
@jrossi 

What would you think about something like the following?

```go
var buildDone = make(chan struct{})

var build = goyek.Define(goyek.Task{
	Name:  "build",
	Action: func(a *goyek.A) {
		t.Parallel()
		// build code...
		close(buildDone)
	}
}

var db_run  = goyek.Define(goyek.Task{
	Name:  "db_run",
	Action: func(a *goyek.A) {
	       findCmd := cmd.NewCmd("db start")
	       statusChan := findCmd.Start() // non-blocking     
	       select {
	       case finalStatus := <-statusChan:
	             // Error something happened with before expected exit
	       case <-buildDone:
                     // build has completed
                     // close db
                     findCmd.Close()
                     return  
	      }
	}
}
var db_migrate  = goyek.Define(goyek.Task{
	Name:  "db_migrate",
	Usage: "",
        Background: false, 
	Deps: goyek.Deps{
		build,
		db_run,
	},
	Action: func(a *goyek.A) {
               cmd.Exec("load_db migrations/*.sql")
	}
}
```

**Comment by jrossi** [comment]
That would work for me and does look a lot simpler to implement then my idea based on goyek.Deps

**Comment by pellared** [comment]
After some exploration, I think it would be better to add `Task.Parallel bool` and `A.RunParallel` from both implementation and user experience perspective.

If the task executor encounters a parallel task then it will batch all ready-to-run parallel tasks together and run them concurrently.

NOTICE: A parallel task must still be able to run a (non-parallel) subtask.

Probably it would be better to first implement `Task.Parallel bool` before trying to implement `A.RunParallel`.
